### PR TITLE
Katherine McConnell

### PR DESCRIPTION
**Pull request by KatherineMcConnell**
Frustrated because things that seemed to have worked before like doctors_by_hospital not working even though they follow the exact same format as other IC's??

Edit: Test world brought in wrong.

### CODE REVIEW

File: lib/doctor.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  attr_reader :name,
+              :specialty,
+              :education,
+              :salary
+  def initialize(employee_info)
+    @name = employee_info[:name]
+    @specialty = employee_info[:specialty]
+    @education = employee_info[:education]
+    @salary = employee_info[:salary]
+  end
```
Comment (mikedao) [inline lib/doctor.rb:11]: Great

File: lib/hospital.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+              :chief_of_surgery,
+              :doctors
+  def initialize(name, chief_of_surgery,doctors)
+    @name = name
+    @chief_of_surgery = chief_of_surgery
+    @doctors = doctors
+  end
+
+  def total_salary
+    @doctors.sum {|doctor| doctor.salary}
```
Comment (mikedao) [inline lib/hospital.rb:12]: Default to the multi line syntax over the single line

File: lib/network.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#     #to_i or to_f and won't let us iterate per google search???
+#     helper_method_for_highest_paid.
+#   end
+# end
+
+
+  def doctors_by_hospital
+    docs_by_hosp = Hash.new {|hash,key| hash[key] = []}
+    @hospitals.each do |hospital|
+      docs_by_hosp[hospital] << doctor.name
```
Comment (mikedao) [inline lib/network.rb:34]: doctor isnt available in scope, this requires a nested iteration approach

File: lib/network.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#     @hospitals.each do |hospital|
+#       hospital.doctors.each do |doctor|
+#          highest_paid << doctor
+#         end
+#       end
+#       highest_paid
+#     end
+#
+#   def highest_paid_doctor
+#     #because 100_000 has underscore and won't let us convert it
```
Comment (mikedao) [inline lib/network.rb:24]: 100_000 is already an integer and does not require conversion

image

File: lib/doctor.rb (hunk 1 above)
Comment (iandouglas) [inline lib/doctor.rb:11]: great job pulling these out of the hash to build your instance variables

File: lib/hospital.rb (hunk 2 above)
Comment (iandouglas) [inline lib/hospital.rb:12]: great use of `.sum` here

File: lib/hospital.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def lowest_paid_doctor
+    @doctors.min_by {|doctor| doctor.salary}.name
+    #this absolutely cannot be convention. But it worked?
```
Comment (iandouglas) [inline lib/hospital.rb:17]: as we talked about in the review, you CAN chain these things but it does get harder to read/debug, but you made it work!

File: lib/hospital.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
+class Hospital
+  attr_reader :name,
+              :chief_of_surgery,
+              :doctors
+  def initialize(name, chief_of_surgery,doctors)
```
Comment (iandouglas) [inline lib/hospital.rb:5]: very very nit-picky, but watch out for the space after the comma between chief/doctors

File: lib/hospital.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    @doctors.sum {|doctor| doctor.salary}
+  end
+
+  def lowest_paid_doctor
+    @doctors.min_by {|doctor| doctor.salary}.name
+    #this absolutely cannot be convention. But it worked?
+  end
+
+  def specialties
+    @doctors.map {|doctor| doctor.specialty}.uniq
```
Comment (iandouglas) [inline lib/hospital.rb:21]: there was no requirement that this was a unique list of specialties, but doesn't hurt to have it here

File: lib/network.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+class Network
+  attr_reader :name,
+              :hospitals
+  def initialize(name)
+    @name = name
+    @hospitals = []
```
Comment (iandouglas) [inline lib/network.rb:6]: good job setting this here

File: lib/network.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+#       end
+#       highest_paid
+#     end
+#
+#   def highest_paid_doctor
+#     #because 100_000 has underscore and won't let us convert it
+#     #to_i or to_f and won't let us iterate per google search???
+#     helper_method_for_highest_paid.
+#   end
+# end
```
Comment (iandouglas) [inline lib/network.rb:28]: the underscore is a thousands-separator but the value should still be considered an Integer class in Ruby, so there should be no conversion necessary. I'm curious what you might have run into here that made this stop working.

File: lib/network.rb (hunk 3 above)
Comment (iandouglas) [inline lib/network.rb:34]: this is a good use of the hash initializer to make sure any missing key is treated as an array as a default so we can just shovel things on here, but where does the `doctor` object come from here? we'd still have to `.map` over each hospital's doctor list or something to get their names here.

File: spec/doctor_spec.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    expect(@meredith).to be_an_instance_of(Doctor)
+  end
+
+  it 'has attributes' do
+
+    expect(@meredith.name).to eq("Meredith Grey")
+    expect(@meredith.specialty).to eq("General Surgery")
+    expect(@meredith.education).to eq("Harvard University")
+    expect(@meredith.salary).to eq(100000)
+  end
```
Comment (iandouglas) [inline spec/doctor_spec.rb:19]: good testing here. The blank lines on line 9 and 14 could be removed

File: spec/hospital_spec.rb [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  it 'exists' do
+
+    expect(@seattle_grace).to be_an_instance_of(Hospital)
+  end
+
+  it 'has attributes' do
+
+    expect(@seattle_grace.name ).to eq("Seattle Grace")
+    expect(@seattle_grace.chief_of_surgery).to eq("Richard Webber")
+    expect(@seattle_grace.doctors).to eq([@meredith, @alex])
```
Comment (iandouglas) [inline spec/hospital_spec.rb:20]: nice use of the instance variables here

File: spec/hospital_spec.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  it 'can sum salaries of doctors' do
+
+    expect(@seattle_grace.total_salary).to eq(190000)
+  end
+
+  it 'can calculate and give name of lowest paid doctor' do
+
+    expect(@seattle_grace.lowest_paid_doctor).to eq("Alex Karev")
```
Comment (iandouglas) [inline spec/hospital_spec.rb:30]: per the review session, you could also use `@alex.name` here too

File: spec/network_spec.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  it 'has attributes' do
+
+    expect(@gsmn.name).to eq("Greater Seattle Medical Network")
+    expect(@gsmn.hospitals).to eq([])
+    @gsmn.add_hospital(@seattle_grace)
+    @gsmn.add_hospital(@grey_sloan)
+
+    expect(@gsmn.hospitals).to eq([@seattle_grace, @grey_sloan])
```
Comment (iandouglas) [inline spec/network_spec.rb:28]: I would split this into two tests, one where you're checking the initialized attributes (single responsibility), and another where you specifically check that calling `.add_hospital` does what you expect (also it's own single responsibility)

**Review by iandouglas — COMMENTED** [review]
Overall, good testing and good structure to the code. Pointed out a few things that I thought were worth mentioning. You got the initial part of i3 done, but the methods there would have bumped this up to a pass.

Score: Fail

**Pull request closed without merging**